### PR TITLE
client: improve the TSO fetching performance

### LINKED ISSUE #3168: The PD client TSO fetching performance rollback
### What did you do?

Prerequisite: the same test environment.

 If you use the old PD client version (90a24a904e34603d4ad89ac1ace05fbcce27cb95) to get the Global TSO, the QPS will look like:

```shell
./bin/pd-tso-bench -pd 172.16.5.39:5299 -v -count 3 -client 1 -c 100

count:474419, max:2, min:0, >1ms:172, >2ms:28, >5ms:0, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
```

If you use the latest PD client version (master), the performance will look like:

```shell
./bin/pd-tso-bench -pd 172.16.5.39:5299 -v -count 3 -client 1 -c 100

count:235362, max:2, min:0, >1ms:8216, >2ms:102, >5ms:0, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
```

### What did you expect to see?

The TSO fetching performance should be similar before and after.

### What did you see instead?

The PD client TSO fetching performance rollback.

### What version of PD are you using (`pd-server -V`)?

master

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

Try to resolve #3168.

### What is changed and how it works?

To modify the TSO fetching mechanism of the PD client, I create multiple goroutines to batch the TSO requests for every `dc-location`. Comparing to the previous version, we don't need to handle all requests in one single goroutine which will reduce the batch performance.

However, the new mechanism is still not as good as the old client (90a24a904e34603d4ad89ac1ace05fbcce27cb95). Everyone, please put your own ideas on how to improve this, thx a lot!

Here are some benchmark results.

#### 1 client 100 concurrencies, 60s duration, requesting the Global TSO with Local TSO on.

This new version:

```shell
Total:
count:1410390, max:61, min:0, >1ms:642773, >2ms:429543, >5ms:196480, >10ms:123962, >30ms:6051 >50ms:149 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:1410390, >1ms:45.57%, >2ms:30.46%, >5ms:13.93%, >10ms:8.79%, >30ms:0.43% >50ms:0.01% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:1556871, max:55, min:0, >1ms:806097, >2ms:421054, >5ms:187638, >10ms:121605, >30ms:4880 >50ms:99 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:1556871, >1ms:51.78%, >2ms:27.04%, >5ms:12.05%, >10ms:7.81%, >30ms:0.31% >50ms:0.01% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:1448181, max:58, min:0, >1ms:669827, >2ms:437955, >5ms:196314, >10ms:127301, >30ms:5307 >50ms:256 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:1448181, >1ms:46.25%, >2ms:30.24%, >5ms:13.56%, >10ms:8.79%, >30ms:0.37% >50ms:0.02% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (1410390+1556871+1448181) * 1/3 * 1/60 = 24530 QPS
```

Previous bad performance version:

```shell
Total:
count:839640, max:75, min:0, >1ms:35269, >2ms:398214, >5ms:228618, >10ms:169338, >30ms:7474 >50ms:628 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:839640, >1ms:4.20%, >2ms:47.43%, >5ms:27.23%, >10ms:20.17%, >30ms:0.89% >50ms:0.07% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:878846, max:52, min:0, >1ms:52291, >2ms:422177, >5ms:220579, >10ms:176560, >30ms:6997 >50ms:75 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:878846, >1ms:5.95%, >2ms:48.04%, >5ms:25.10%, >10ms:20.09%, >30ms:0.80% >50ms:0.01% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:846975, max:103, min:0, >1ms:45928, >2ms:410730, >5ms:208726, >10ms:172620, >30ms:8264 >50ms:522 >100ms:49 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:846975, >1ms:5.42%, >2ms:48.49%, >5ms:24.64%, >10ms:20.38%, >30ms:0.98% >50ms:0.06% >100ms:0.01% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (839640+878846+846975) * 1/3 * 1/60 = 14252 QPS
```

Old good performance version:

```shell
Total:
count:2527759, max:23, min:0, >1ms:1843069, >2ms:442736, >5ms:177431, >10ms:30535, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:2527759, >1ms:72.91%, >2ms:17.51%, >5ms:7.02%, >10ms:1.21%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:2542472, max:40, min:0, >1ms:1876511, >2ms:436089, >5ms:161975, >10ms:36236, >30ms:151 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:2542472, >1ms:73.81%, >2ms:17.15%, >5ms:6.37%, >10ms:1.43%, >30ms:0.01% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:2578043, max:27, min:0, >1ms:1893938, >2ms:429821, >5ms:183430, >10ms:34210, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:2578043, >1ms:73.46%, >2ms:16.67%, >5ms:7.12%, >10ms:1.33%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (2527759+2542472+2578043) * 1/3 * 1/60 = 42490 QPS
```

#### 1 client 100 concurrencies, 60s duration, requesting the Global TSO with Local TSO off.

This new version:

```shell
Total:
count:6692723, max:46, min:0, >1ms:271060, >2ms:193284, >5ms:108767, >10ms:51638, >30ms:1175 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:6692723, >1ms:4.05%, >2ms:2.89%, >5ms:1.63%, >10ms:0.77%, >30ms:0.02% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:7129421, max:43, min:0, >1ms:263149, >2ms:174561, >5ms:107454, >10ms:54275, >30ms:957 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:7129421, >1ms:3.69%, >2ms:2.45%, >5ms:1.51%, >10ms:0.76%, >30ms:0.01% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:6974326, max:49, min:0, >1ms:273861, >2ms:188648, >5ms:113275, >10ms:52515, >30ms:1124 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:6974326, >1ms:3.93%, >2ms:2.70%, >5ms:1.62%, >10ms:0.75%, >30ms:0.02% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (6692723+7129421+6974326) * 1/3 * 1/60 = 115535 QPS
```

Previous bad performance version:

```shell
Total:
count:3689620, max:58, min:0, >1ms:524382, >2ms:213206, >5ms:164852, >10ms:90861, >30ms:1478 >50ms:100 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:3689620, >1ms:14.21%, >2ms:5.78%, >5ms:4.47%, >10ms:2.46%, >30ms:0.04% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:3793496, max:61, min:0, >1ms:458708, >2ms:208965, >5ms:151782, >10ms:92445, >30ms:1799 >50ms:65 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:3793496, >1ms:12.09%, >2ms:5.51%, >5ms:4.00%, >10ms:2.44%, >30ms:0.05% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:3923527, max:44, min:0, >1ms:467380, >2ms:209478, >5ms:154138, >10ms:87633, >30ms:1318 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:3923527, >1ms:11.91%, >2ms:5.34%, >5ms:3.93%, >10ms:2.23%, >30ms:0.03% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (3689620+3793496+3923527) * 1/3 * 1/60 = 63370 QPS
```

Old good performance version:

```shell
Total:
count:10331449, max:28, min:0, >1ms:173317, >2ms:119953, >5ms:56878, >10ms:15353, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:10331449, >1ms:1.68%, >2ms:1.16%, >5ms:0.55%, >10ms:0.15%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:10607982, max:24, min:0, >1ms:155515, >2ms:112757, >5ms:65143, >10ms:13927, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:10607982, >1ms:1.47%, >2ms:1.06%, >5ms:0.61%, >10ms:0.13%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:10756544, max:26, min:0, >1ms:158814, >2ms:117927, >5ms:59726, >10ms:13618, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:10756544, >1ms:1.48%, >2ms:1.10%, >5ms:0.56%, >10ms:0.13%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (10331449+10607982+10756544) * 1/3 * 1/60 = 176088 QPS
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

Related changes

- PR #3045

### CODE REVIEW

**Comment by nolouch** [comment]
How about local TSO off?

File: client/client.go [hunk 1]
```diff
@@ -283,138 +287,116 @@ func (c *client) tsLoop() {
...
 }
 
-func (c *client) checkTSOStreams(streams map[string]pdpb.PD_TsoClient) bool {
-	dcLocations := c.getDCLocations()
-	if len(streams) != len(dcLocations) {
+func (c *client) checkTSODispatcher(dcLocation string) bool {
+	c.tsoDispatcher.RLock()
+	defer c.tsoDispatcher.RUnlock()
+	tsoChan, exist := c.tsoDispatcher.ch[dcLocation]
+	if !exist || tsoChan == nil {
```
Comment (rleungx) [inline client/client.go:387]: When will the tsoChan become nil?

File: client/client.go [hunk 2]
```diff
@@ -283,138 +287,116 @@ func (c *client) tsLoop() {
...
+					)
+					for {
+						ctx, cancel := context.WithCancel(loopCtx)
+						if stream == nil {
+							go c.checkStreamTimeout(loopCtx, cancel, createdCh)
+							stream, err = pdpb.NewPDClient(c.connMu.clientConns[c.connMu.allocators[dc]]).Tso(ctx)
+							if stream != nil {
+								createdCh <- struct{}{}
+							}
+							if err != nil {
```
Comment (rleungx) [inline client/client.go:315]: Why not check the error first?

File: client/client.go [hunk 3]
```diff
@@ -283,138 +287,116 @@ func (c *client) tsLoop() {
...
+						err       error
+						stream    pdpb.PD_TsoClient
+						opts      []opentracing.StartSpanOption
+						requests  = make([]*tsoRequest, maxMergeTSORequests+1)
+						createdCh = make(chan struct{})
+					)
+					for {
+						ctx, cancel := context.WithCancel(loopCtx)
+						if stream == nil {
+							go c.checkStreamTimeout(loopCtx, cancel, createdCh)
```
Comment (rleungx) [inline client/client.go:310]: Do we need to use ctx instead of loopCtx?

**Comment by JmPotato** [comment]
> How about local TSO off?

The QPS will both increase, however the performance gap ratio is still the same.

**Review by Yisaer — COMMENTED** [review]
Maybe we should add a bench test first?

File: client/client.go (hunk 1 above)
Comment (JmPotato) [reply client/client.go:387]: According to the existing code, this will not happen.

**Comment by ZenoTan** [comment]
If global TSO only, I think it should be the same as before. Why it is faster?

File: client/client.go [hunk 4]
```diff
@@ -283,138 +287,113 @@ func (c *client) tsLoop() {
...
+								return
+							default:
+							}
+							log.Error("[pd] getTS error", zap.String("dc-location", dc), errs.ZapError(errs.ErrClientGetTSO, err))
+							c.ScheduleCheckLeader()
+							cancel()
+							stream = nil
+						}
+					}
+				}(dcLocation, c.tsoDispatcher.ch[dcLocation])
```
Comment (nolouch) [inline client/client.go:374]: Does the outer loop need to wait the inner loop's goroutine?

Comment (JmPotato) [reply client/client.go:374]: The outer loop is used to create the goroutine for a new `dc-location`, so it doesn't need to wait for the inner goroutines. The `loopCtx` controls the lifetimes of the inner goroutines.

Comment (nolouch) [reply client/client.go:374]: Does `tsLoop` always run the loop started in `line 290`?

Comment (JmPotato) [reply client/client.go:374]: Yes, it will keep looping to check whether we have new DC joined.

**Comment by JmPotato** [comment]
Here are some new results, PTAL @nolouch.

#### 1 client 100 concurrencies, 60s duration, requesting the Global TSO with Local TSO on.

This new version:

```shell
Total:
count:4036344, max:22, min:0, >1ms:3491867, >2ms:294228, >5ms:5392, >10ms:1370, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:4036344, >1ms:86.51%, >2ms:7.29%, >5ms:0.13%, >10ms:0.03%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:3772737, max:14, min:0, >1ms:3111424, >2ms:366387, >5ms:8273, >10ms:405, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:3772737, >1ms:82.47%, >2ms:9.71%, >5ms:0.22%, >10ms:0.01%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:3939770, max:17, min:0, >1ms:3278111, >2ms:353056, >5ms:5046, >10ms:796, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:3939770, >1ms:83.21%, >2ms:8.96%, >5ms:0.13%, >10ms:0.02%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (4036344 + 3772737 + 3939770) * 1/3 * 1/60 = 65271 QPS
```

Previous bad performance version:

```shell
Total:
count:2576561, max:49, min:0, >1ms:1444939, >2ms:1023563, >5ms:50310, >10ms:9455, >30ms:699 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:2576561, >1ms:56.08%, >2ms:39.73%, >5ms:1.95%, >10ms:0.37%, >30ms:0.03% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:2725110, max:93, min:0, >1ms:1665302, >2ms:950362, >5ms:45013, >10ms:6950, >30ms:862 >50ms:315 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:2725110, >1ms:61.11%, >2ms:34.87%, >5ms:1.65%, >10ms:0.26%, >30ms:0.03% >50ms:0.01% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:2735030, max:29, min:0, >1ms:1634162, >2ms:993554, >5ms:48239, >10ms:4533, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:2735030, >1ms:59.75%, >2ms:36.33%, >5ms:1.76%, >10ms:0.17%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (2576561 + 2725110 + 2735030) * 1/3 * 1/60 = 44648 QPS
```

Old good performance version:

```shell
Total:
count:4140666, max:18, min:0, >1ms:3708333, >2ms:140239, >5ms:1439, >10ms:1171, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:4140666, >1ms:89.56%, >2ms:3.39%, >5ms:0.03%, >10ms:0.03%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:4162000, max:25, min:0, >1ms:3779213, >2ms:123846, >5ms:2817, >10ms:300, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:4162000, >1ms:90.80%, >2ms:2.98%, >5ms:0.07%, >10ms:0.01%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:4082852, max:26, min:0, >1ms:3711833, >2ms:152472, >5ms:1405, >10ms:1639, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:4082852, >1ms:90.91%, >2ms:3.73%, >5ms:0.03%, >10ms:0.04%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%


Average: (4140666 + 4162000 + 4082852) * 1/3 * 1/60 = 68808 QPS
```

#### 1 client 100 concurrencies, 60s duration, requesting the Global TSO with Local TSO off.

This new version:

```shell
Total:
count:23839003, max:11, min:0, >1ms:36623, >2ms:20377, >5ms:2057, >10ms:21, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:23839003, >1ms:0.15%, >2ms:0.09%, >5ms:0.01%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:24152263, max:11, min:0, >1ms:26809, >2ms:14783, >5ms:776, >10ms:37, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:24152263, >1ms:0.11%, >2ms:0.06%, >5ms:0.00%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:23734834, max:14, min:0, >1ms:36441, >2ms:20559, >5ms:1846, >10ms:51, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:23734834, >1ms:0.15%, >2ms:0.09%, >5ms:0.01%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (23839003 + 24152263 + 23734834) * 1/3 * 1/60 = 398478 QPS
```

Previous bad performance version:

```shell
Total:
count:12079401, max:13, min:0, >1ms:346864, >2ms:141912, >5ms:16352, >10ms:493, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:12079401, >1ms:2.87%, >2ms:1.17%, >5ms:0.14%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:11953634, max:19, min:0, >1ms:313129, >2ms:150919, >5ms:19237, >10ms:683, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:11953634, >1ms:2.62%, >2ms:1.26%, >5ms:0.16%, >10ms:0.01%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:12767788, max:11, min:0, >1ms:300035, >2ms:146473, >5ms:20260, >10ms:253, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:12767788, >1ms:2.35%, >2ms:1.15%, >5ms:0.16%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (12079401 + 11953634 + 12767788) * 1/3 * 1/60 = 204449 QPS
```

Old good performance version:

```shell
Total:
count:21040046, max:39, min:0, >1ms:77983, >2ms:16566, >5ms:225, >10ms:0, >30ms:100 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:21040046, >1ms:0.37%, >2ms:0.08%, >5ms:0.00%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:20361055, max:39, min:0, >1ms:96215, >2ms:37119, >5ms:1187, >10ms:101, >30ms:100 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:20361055, >1ms:0.47%, >2ms:0.18%, >5ms:0.01%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Total:
count:20861289, max:7, min:0, >1ms:91560, >2ms:17761, >5ms:2, >10ms:0, >30ms:0 >50ms:0 >100ms:0 >200ms:0 >400ms:0 >800ms:0 >1s:0
count:20861289, >1ms:0.44%, >2ms:0.09%, >5ms:0.00%, >10ms:0.00%, >30ms:0.00% >50ms:0.00% >100ms:0.00% >200ms:0.00% >400ms:0.00% >800ms:0.00% >1s:0.00%

Average: (21040046 + 20361055 + 20861289) * 1/3 * 1/60 = 345902 QPS
```

File: client/client.go [hunk 5]
```diff
@@ -558,11 +532,39 @@ func (c *client) GetLocalTSAsync(ctx context.Context, dcLocation string) TSFutur
...
+	c.dispatchRequest(dcLocation, req)
 	return req
 }
 
+func (c *client) waitForDispatcher() {
+	for {
+		if c.getDispatcherSize() != 0 {
+			break
+		}
+		log.Info("[pd] tso dispatcher is not ready, wait for a while")
```
[the remark this answers is not in the document]
Comment (JmPotato) [reply client/client.go:551]: Technically not, because this situation only happens when the client has just started.

**Comment by JmPotato** [comment]
@nolouch @Yisaer @rleungx @disksing PTAL

File: client/client.go [hunk 6]
```diff
@@ -157,18 +163,13 @@ var (
...
-		ch map[string]chan deadline
-	}
+	tsDeadline sync.Map // Same as map[string]chan deadline
 	// dc-location -> int64
-	lastTSO struct {
-		sync.Mutex
-		physical map[string]int64
-		logical  map[string]int64
-	}
+	lastTSMap sync.Map // Same as map[string]*lastTSO
```
Comment (Yisaer) [inline client/client.go:172]: The comment here seems not right.

**Review by Yisaer — COMMENTED** [review]
rest LGTM

File: client/client.go (hunk 6 above)
Comment (JmPotato) [reply client/client.go]: Updated.

File: client/client.go [hunk 7]
```diff
@@ -288,138 +288,111 @@ func (c *client) tsLoop() {
...
-				}
-				wg.Add(1)
-				go func(dc string, reqs []*tsoRequest) {
-					defer wg.Done()
-					done := make(chan struct{})
-					dl := deadline{
-						timer:  time.After(c.timeout),
-						done:   done,
-						cancel: cancel,
+		c.allocators.Range(func(dcLocationKey, _ interface{}) bool {
```
Comment (nolouch) [inline client/client.go:294]: Does this affect the performance?  before no need mutex.

Comment (JmPotato) [reply client/client.go:294]: You can check the [source codes](https://golang.org/src/sync/map.go#L308) of the `Range` method. It mostly iterates over a read-only map data, which won't block other writing or reading operations and doesn't involve a `Mutex` in most cases.

Comment (nolouch) [reply client/client.go:294]: got it!

**Review by nolouch — APPROVED** [review]
lgtm

**Review by Yisaer — APPROVED** [review]
LGTM

**Pull request merged**